app/src/main/java/com/google/samples/apps/sunflower/viewmodels/PlantDetailViewModel.kt

```diff
@@ -66,10 +66,14 @@ class PlantDetailViewModel(
     fun toggleAddOrRemove(view: View) {
         if (isPlanted.value == true) {
             removePlantFromGarden()
-            notifyToggleAddOrRemove.postValue(false)
+            showSnackbarMessage(R.string.removed_plant_from_garden)
         } else {
             addPlantToGarden()
-            notifyToggleAddOrRemove.postValue(true)
+            showSnackbarMessage(R.string.added_plant_to_garden)
         }
     }
+
+    private fun showSnackbarMessage(@StringRes message: Int?) {
+        snackbarText.value = message
```

Comment:
Nice2have: I prefer to `postValue`

Reply:
@XinyueZ I see the usage in https://github.com/googlesamples/android-architecture and copy it. Do you mean toggleAddOrRemove in ViewModel need support call from background thread?

Reply:
@XinyueZ I see the usage in https://github.com/googlesamples/android-architecture and copy it. Do you mean toggleAddOrRemove in ViewModel need support call from background thread?

Reply:
any reason why you prefer `postValue`?

Reply:
Like choose of  `postInvalidat` and `invalidate`. I am hesitate here whether
`showSnackerMessage` can be tested pass through. However it's only my
suggestion which is nice2have. Try to write ui.tests which could associate
with this method, if pass, I can accept  setValue as well.

Reply:
@williamwue try to write a test *on* `PlantDetailViewModel` with `showSnackbarMessage`, if it can pass, take `setValue` afterward, otherwise, `postValue` .

Reply:
it works actually. just need to use `InstantTaskExecutorRule` by adding `androidTestImplementation "android.arch.core:core-testing:1.1.1"`

```
@get:Rule
var rule: TestRule = InstantTaskExecutorRule()

@Test
fun testSnackbarMessage_AddedPlantToGardenString() {
    assertNotEquals(viewModel.snackbarText.value, R.string.added_plant_to_garden)
    viewModel.showSnackbarMessage(R.string.added_plant_to_garden)
    assertEquals(viewModel.snackbarText.value, R.string.added_plant_to_garden)
}
```

https://developer.android.com/reference/android/arch/core/executor/testing/InstantTaskExecutorRule

Reply:
@andhie OK, thanks for sharing, accept and nice2have this TEST of the feature of DELETE.

Reply:
@XinyueZ @andhie OK, I got it and will add test for it.